File: lab-3-dynamodb/dynamoDBCSharpLab/Lab3/PatientReportLinker.cs
// Copyright 2017 Amazon Web Services, Inc. or its affiliates. All rights reserved.

using Amazon.DynamoDBv2;
using Amazon.DynamoDBv2.Model;
using Newtonsoft.Json;
using System.Collections.Generic;
using System.Diagnostics;

namespace Lab3
{
    // The PatientReportLinker class updates DynamoDB items with the corresponding
    // link to a patient's report on S3
    public static class PatientReportLinker
    {
        public static readonly string InfectionsTableName = InfectionsTableCreator.InfectionsTableName;
        public static readonly string PatientReportPrefix = Utils.PatientReportPrefix;
        public static readonly string S3BucketName = Utils.LabS3BucketName;
        public static readonly string S3BucketRegion = Utils.LabS3BucketRegion;
        public static AmazonDynamoDBClient dynamoDBClient = null;

        public static void main()
        {
            Init();
            LinkPatientReport();
        }

        private static void LinkPatientReport()
        {
            string reportUrl = null;
            string objectKey = null;

            // Sample reports exist for patient ids 1, 2, 3
            for (int i = 1; i < 4; i++)
            {
                objectKey = PatientReportPrefix + i + ".txt";
                reportUrl = "https://s3-" + S3BucketRegion + ".amazonaws.com/" + S3BucketName + "/" + objectKey;
                UpdateItemWithLink(("" + i), reportUrl);
            }
        }

        private static void UpdateItemWithLink(string patientId, string reportUrl)
        {
            Debug.WriteLine("Updating item patientId: {0}, reportURL: {1}", patientId, reportUrl);

            UpdateItemRequest requestUpdate = new UpdateItemRequest
            {
                TableName = InfectionsTableName,
                Key = new Dictionary<string, AttributeValue>()
                        {
                            { "PatientID", new AttributeValue { S = patientId } }
                        },
                ExpressionAttributeNames = new Dictionary<string, string>()
                        {
                            { "#purl", "PatientReportUrl" },
                        },
                ExpressionAttributeValues = new Dictionary<string, AttributeValue>()
                        {
                             { ":val1", new AttributeValue {S = reportUrl} },
                        },
                // This expression does the following:
                // Adds a new attribute to the item
                UpdateExpression = "SET #purl = :val1"
            };

            UpdateItemResponse responseUpdate = UpdateItem(requestUpdate);
            Debug.WriteLine("Printing item after adding attribute:");

            string jsonDisplayText = ConvertToJSON(responseUpdate);
            Debug.WriteLine("Display item.");
            Debug.WriteLine(jsonDisplayText);
        }

        private static void Init()
        {
            dynamoDBClient = new AmazonDynamoDBClient();
        }

        /**
         * Update the item in the table
         *
         * @param requestUpdate     Update Item Request
         * @return                  Order
         */
        private static UpdateItemResponse UpdateItem(UpdateItemRequest requestUpdate)
        {
            // STUDENT TODO 5: Replace the solution with your own code
            return Solution.UpdateItem(dynamoDBClient, requestUpdate);
        }

        /**
         * Retrieve the item from the UpdateItemOutcome object and print
         * its contents in JSON format
         *
         * @param response     Order
         * @return             Order in JSON format
         */
        private static string ConvertToJSON(UpdateItemResponse response)
        {
            // STUDENT TODO 6: Replace the solution with your own code
            return Solution.ConvertToJSON(response);
        }
    }
}

File: lab-3-dynamodb/dynamoDBJavaLab/src/PatientReportLinker.java
// Copyright 2017 Amazon Web Services, Inc. or its affiliates. All rights reserved.

import com.amazonaws.regions.Region;
import com.amazonaws.services.dynamodbv2.AmazonDynamoDBClient;
import com.amazonaws.services.dynamodbv2.document.DynamoDB;
import com.amazonaws.services.dynamodbv2.document.UpdateItemOutcome;

// The PatientReportLinker class updates DynamoDB items with the corresponding link to a patient's report on S3.
public class PatientReportLinker {

  public static final String INFECTIONS_TABLE_NAME = InfectionsTableCreator.INFECTIONS_TABLE_NAME;
  public static final String PATIENT_REPORT_PREFIX = Utils.PATIENT_REPORT_PREFIX;
  public static final String S3_BUCKET_NAME = Utils.LAB_S3_BUCKET_NAME;
  public static final String S3_BUCKET_REGION = Utils.LAB_S3_BUCKET_REGION;

  public static final Region REGION = Utils.getRegion();

  public static DynamoDB dynamoDB = null;
  private static AmazonDynamoDBClient dynamoDBClient = null;

  public static void main(String[] args) throws Exception {

    // Instantiate DynamoDB client and object
    dynamoDBClient = new AmazonDynamoDBClient();
    dynamoDBClient.setRegion(REGION);
    dynamoDB = new DynamoDB(dynamoDBClient);

    String reportUrl = null;
    String objectKey = null;

    // Sample reports exist for patient ids 1, 2, 3
    for (int i = 1; i < 4; i++) {
      objectKey = PATIENT_REPORT_PREFIX + i + ".txt";

      // Construct the URL for the patient report
      reportUrl =
          "https://s3-" + S3_BUCKET_REGION + ".amazonaws.com/" + S3_BUCKET_NAME + "/" + objectKey;

      System.out.printf("Updating item for patientId: %d %n", i);

      // Update the DynamoDB item with the link
      UpdateItemOutcome outcome =
          updateItemWithLink(dynamoDB, INFECTIONS_TABLE_NAME, ("" + i), reportUrl);
      System.out.println("\nPrinting item after adding attribute:");

      // Print item in JSON format
      System.out.println(outcome.getItem().toJSONPretty());
      System.out.println("-------------------------------------");
    }
  }

  /**
   * Update the item for the patient id with an attribute called PatientReportUrl
   *
   * @param dynamoDB    Instance of DynamoDB class
   * @param tableName   Table name
   * @param patientId   Patient ID
   * @param reportUrl   Report ID
   * @return            Query results
   */
  private static UpdateItemOutcome updateItemWithLink(
      DynamoDB dynamoDB, String tableName, String patientId, String reportUrl) {
    // STUDENT TODO 4: Replace the solution with your own code
    return Solution.updateItemWithLink(dynamoDB, tableName, patientId, reportUrl);
  }
}
